criu/files-reg.c

```diff
@@ -1380,12 +1387,276 @@ static bool should_check_size(int flags)
+			program_header->p_type != PT_NOTE)
+		program_header++;
+
+	if (!num_iterations || program_header >= program_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	note_header = (Elf64_Nhdr *) (program_header->p_offset + (char *) file_header);
+	if (note_header <= (Elf64_Nhdr *) file_header) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	note_header_end = (Elf64_Nhdr *) (file_header_end - sizeof(Elf64_Nhdr));
+	num_iterations = 500;
+
+	/* The note type for the build-id is NT_GNU_BUILD_ID. */
+	while (num_iterations-- && note_header <= note_header_end &&
+			note_header->n_type != NT_GNU_BUILD_ID)
+		note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+						note_header->n_namesz + note_header->n_descsz);
+
+	if (!num_iterations || note_header >= note_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	if (note_header->n_descsz <= 0 || note_header->n_descsz > 512) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	size = note_header->n_descsz;
+	note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+					note_header->n_namesz);
+	note_header_end = (Elf64_Nhdr *) (file_header_end - size);
+	if (note_header <= (Elf64_Nhdr *) file_header || note_header > note_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	*build_id = (unsigned char *) xmalloc(size);
+	if (!*build_id) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	memcpy(*build_id, (void *) note_header, size);
+
+	munmap(file_header, fd_status->st_size);
+	return size;
+}
+
+/*
+ * Does the actual work of finding the build-id of the file, if it exists.
+ * http://ftp.openwatcom.org/devel/docs/elf-64-gen.pdf
+ */
+static int get_build_id(const int fd, const struct stat *fd_status,
+				unsigned char **build_id)
+{
+	char *start_addr;
+
+	if (fd_status->st_size < 5)
+		return -1;
+	start_addr = (char *) mmap(0, fd_status->st_size,
+					PROT_READ, MAP_PRIVATE, fd, 0);
+	if (start_addr == MAP_FAILED)
+		return -1;
+
+	if (start_addr[0] != 0x7f || start_addr[1] != 0x45 ||
+		start_addr[2] != 0x4c || start_addr[3] != 0x46) {
+		munmap(start_addr, fd_status->st_size);
+		return -1;
+	}
+
+	if (start_addr[4] == ELFCLASS64)
+		return get_build_id_64((Elf64_Ehdr *) start_addr, fd_status, build_id);
+	else if (start_addr[4] == ELFCLASS32)
+		return get_build_id_32((Elf32_Ehdr *) start_addr, fd_status, build_id);
+	else
+		return -1;
+}
+
+/*
+ * Finds and stores the build-id of a file, if it exists, so that it can be validated
+ * while restoring.
+ */
+static bool store_validation_data_build_id(RegFileEntry *rfe, int lfd)
+{
+	unsigned char *build_id = NULL;
+	int build_id_size, i;
+	char buf[32];
+	int fd;
+	struct stat st;
+
+	snprintf(buf, sizeof(buf), "/proc/self/fd/%d", lfd);
+	fd = open(buf, O_RDONLY);
+	if (fd < 0)
+		return false;
+	if (fstat(fd, &st) < 0) {
```

Comment:
fd_param contains stat data for this fd so we probably can avoid this fstat here.